File: examples/python/file_transfer.py
#!/usr/bin/env python3
"""UniNet — sending a file to another machine.

    python3 file_transfer.py receive [output-dir]     # terminal 1
    python3 file_transfer.py send <path> [more...]    # terminal 2

No address is configured. The sender finds the receiver and streams the file
with progress at both ends.
"""
import os
import sys
import time

import uninet

SUBJECT = "files"


def human(n: float) -> str:
    for unit in ("B", "KB", "MB", "GB"):
        if n < 1024 or unit == "GB":
            return f"{n:.1f} {unit}" if unit != "B" else f"{int(n)} B"
        n /= 1024
    return f"{n:.1f} GB"


def receive(outdir: str) -> int:
    os.makedirs(outdir, exist_ok=True)
    net = uninet.join("File Receiver", role="viewer", app="uninet-examples")
    blob = uninet.Blob(net, SUBJECT)

    def on_progress(info, done):
        pct = 100.0 * done / info.size if info.size else 100.0
        print(f"\r  {info.name}: {pct:5.1f}%  {human(done)}/{human(info.size)}",
              end="", flush=True)

    def on_received(info, data):
        # Never trust a remote-supplied filename: basename() stops "../../etc/x"
        # from escaping the output directory.
        safe = os.path.basename(info.name) or "unnamed"
        path = os.path.join(outdir, safe)
        with open(path, "wb") as fh:
            fh.write(data)
        print(f"\r  {info.name}: saved to {path} ({human(len(data))})          ")
        if info.meta and info.meta.get("sha256"):
            import hashlib
            got = hashlib.sha256(data).hexdigest()
            ok = "OK" if got == info.meta["sha256"] else "MISMATCH"
            print(f"    sha256 {ok}")

    blob.on_progress(on_progress)
    blob.on_received(on_received)
    blob.on_failed(lambda info, why: print(f"\r  {info.name} failed: {why}"))

    print(net.describe())
    print(f"Saving incoming files to {outdir}. Ctrl+C to stop.\n")
    try:
        while True:
            time.sleep(0.5)
    except KeyboardInterrupt:
        print("\nstopped")
    return 0


def send(paths) -> int:
    import hashlib

    net = uninet.join("File Sender", role="server", app="uninet-examples")
    blob = uninet.Blob(net, SUBJECT)

    print(net.describe())
    print("Looking for a receiver…")
    for _ in range(200):
        if net.peers():
            break
        time.sleep(0.1)
    if not net.peers():
        print("No receiver found. Start 'file_transfer.py receive' first.")
        return 1
    print(f"Found {net.peers()[0].name} at {net.peers()[0].endpoint}\n")

    for path in paths:
        if not os.path.isfile(path):
            print(f"  skipping {path}: not a file")
            continue
        with open(path, "rb") as fh:
            digest = hashlib.sha256(fh.read()).hexdigest()
        size = os.path.getsize(path)
        print(f"  sending {os.path.basename(path)} ({human(size)})")
        blob.send_file(path, meta={"sha256": digest, "size": size})

    # Let the chunks drain before the session goes away.
    time.sleep(3)
    print("done")
    return 0


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else "receive"
    if mode == "send" and len(sys.argv) > 2:
        sys.exit(send(sys.argv[2:]))
    elif mode == "receive":
        sys.exit(receive(sys.argv[2] if len(sys.argv) > 2 else "./received"))
    else:
        print(__doc__)
        sys.exit(2)

File: examples/file_transfer.cpp
// UniNet: sending a file to another machine, in C++.
//
//     uninet-file-transfer receive [output-dir]     # terminal 1
//     uninet-file-transfer send <path> [more...]    # terminal 2
//
// Or on two machines, one running each. No address is configured: the sender
// finds the receiver and streams the file, with progress at both ends.
//
// publish() would be the wrong tool here, a file has to be chunked, and the
// receiver wants to know it is 40% through. Blob does both.
#include "uninet/blob.h"
#include "uninet/session.h"

#include <atomic>
#include <csignal>
#include <chrono>
#include <cstdio>
#include <filesystem>
#include <cstring>
#include <fstream>
#include <string>
#include <thread>
#include <vector>

namespace {

constexpr const char* kSubject = "files";

std::string human(double n) {
    const char* units[] = {"B", "KB", "MB", "GB"};
    int u = 0;
    while (n >= 1024.0 && u < 3) { n /= 1024.0; ++u; }
    char buf[48];
    std::snprintf(buf, sizeof buf, u == 0 ? "%.0f %s" : "%.1f %s", n, units[u]);
    return buf;
}

std::string basename_of(const std::string& path) {
    const size_t slash = path.find_last_of("/\\");
    return slash == std::string::npos ? path : path.substr(slash + 1);
}

std::atomic<bool> g_stop{false};
void on_signal(int) { g_stop.store(true); }

int receive(const std::string& outdir) {
    // Before joining, not when the first file lands: the directory defaults to
    // ./received, nothing creates it, and ofstream on a path in a directory
    // that does not exist simply fails - so a transfer that both ends reported
    // as complete left no file anywhere.
    std::error_code ec;
    std::filesystem::create_directories(outdir, ec);
    if (ec) {
        std::printf("Cannot create %s: %s\n", outdir.c_str(), ec.message().c_str());
        return 1;
    }

    auto net = uninet::Session::join("File Receiver", [] {
        uninet::SessionConfig c;
        c.role = "viewer";
        c.app  = "uninet-examples";
        return c;
    }());

    uninet::Blob blob(*net, kSubject);

    blob.on_progress([](const uninet::BlobInfo& info, size_t done) {
        const double pct = info.size ? 100.0 * double(done) / double(info.size) : 100.0;
        std::printf("\r  %s: %5.1f%%  %s/%s", info.name.c_str(), pct,
                    human(double(done)).c_str(), human(double(info.size)).c_str());
        std::fflush(stdout);
    });

    blob.on_received([outdir](const uninet::BlobInfo& info, const uninet::Bytes& data) {
        // Never trust a remote-supplied filename: taking the basename stops
        // "../../etc/passwd" from escaping the output directory.
        const std::string safe = basename_of(info.name).empty() ? "unnamed"
                                                                : basename_of(info.name);
        const std::string path = outdir + "/" + safe;
        std::ofstream out(path, std::ios::binary);
        if (!out) {
            std::printf("\r  %s: could not write %s\n", info.name.c_str(), path.c_str());
            return;
        }
        out.write(reinterpret_cast<const char*>(data.data()), std::streamsize(data.size()));
        std::printf("\r  %s: saved to %s (%s)          \n",
                    info.name.c_str(), path.c_str(), human(double(data.size())).c_str());
    });

    blob.on_failed([](const uninet::BlobInfo& info, const std::string& why) {
        std::printf("\r  %s failed: %s\n", info.name.c_str(), why.c_str());
    });

    std::printf("%s\n", net->describe().c_str());
    std::printf("Saving incoming files to %s. Ctrl+C to stop.\n\n", outdir.c_str());

    std::signal(SIGINT, on_signal);
    std::signal(SIGTERM, on_signal);
    while (!g_stop.load()) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    return 0;
}

int send(const std::vector<std::string>& paths) {
    auto net = uninet::Session::join("File Sender", [] {
        uninet::SessionConfig c;
        c.role = "server";
        c.app  = "uninet-examples";
        return c;
    }());

    uninet::Blob blob(*net, kSubject);

    std::printf("%s\nLooking for a receiver...\n", net->describe().c_str());
    for (int i = 0; i < 200 && net->peers().empty(); ++i)
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    if (net->peers().empty()) {
        std::printf("No receiver found. Start 'uninet-file-transfer receive' first.\n");
        return 1;
    }
    std::printf("Found %s at %s\n\n", net->peers()[0].name.c_str(),
                net->peers()[0].endpoint().c_str());

    for (const auto& path : paths) {
        std::ifstream probe(path, std::ios::binary | std::ios::ate);
        if (!probe) { std::printf("  skipping %s: cannot read it\n", path.c_str()); continue; }
        std::printf("  sending %s (%s)\n", basename_of(path).c_str(),
                    human(double(probe.tellg())).c_str());
        if (blob.send_file(path).empty())
            std::printf("    failed to start the transfer\n");
    }

    // Let the chunks drain before the session, and its network thread: goes away.
    std::this_thread::sleep_for(std::chrono::seconds(3));
    std::printf("done\n");
    return 0;
}

}  // namespace

int main(int argc, char** argv) {
    const std::string mode = argc > 1 ? argv[1] : "receive";
    if (mode == "receive") {
        return receive(argc > 2 ? argv[2] : "./received");
    }
    if (mode == "send" && argc > 2) {
        return send(std::vector<std::string>(argv + 2, argv + argc));
    }
    std::printf("uninet-file-transfer: send a file to another machine, no address needed\n\n"
                "  uninet-file-transfer receive [output-dir]\n"
                "  uninet-file-transfer send <path> [more...]\n");
    return 2;
}
